specifications/transactions.md

```diff
@@ -0,0 +1,619 @@
+    ChainID        (uint32)
+    ConsumedTxIdx  (uint32)
+    ConsumedTxHash (32 byte array)
+}
+```
+
+ *  TxIn
+    - TxInLinker:
+      specifies the TxInLinker object
+    - Signature:
+      specifies the signature of the object
+ *  TxInLinker
+    - TxInPreImage:
+      specifies the TxInPreImage object
+    - TxHash:
+      specifies the transaction hash of the object
+ *  TxInPreImage
+    - ChainID:
+      specifies the ChainID for this object;
+      **must** be nonzero
+    - ConsumedTxIdx:
+      specifies the TxOutIdx of the UTXO being consumed
+    - ConsumedTxHash:
+      specifies the TxHash of the UTXO being consumed
+
+### Deposit
+
+A Deposit is a special type of TxIn:
+it specifies value which has been transferred to AliceNet.
+Deposits may only be consumed.
+It is important to note that a
+**deposit's validity must come from outside of AliceNet**;
+this is a necessary part of the system.
+
+Each deposit type will be contained within their own separate
+subtrees of the `StateTrie`;
+the location is specified by `key` (32 byte value)
+and `number` (`uint256` value starting at 1).
+
+### Tx
+
+The transaction object is central to AliceNet.
+The primary components of a transaction are
+the list of consumed UTXOs (specified as an array of TxIn objects)
+and the list of new UTXOs (specified as an array of TxOut objects).
+It also contains the type and data information
+necessary for transaction validation.
+A Fee object is included for completeness and is required
+for validation;
+it is used implicitly in the TxHash computation.
+
+```
+Tx {
+    Type (uint32)
+    Data (byte array)
+    Fee  (uint256)
+    Vin  (TxIn array)
+    Vout (TxOut array)
+}
+```
+
+ *  Tx
+    - Type:
+      specifies the type of the transaction object
+    - Data:
+      specifies additional data required for transaction processing
+    - Fee:
+      specifies the total transaction fee
+    - Vin:
+      specifies the list of consumed UTXOs;
+      **must** be nonempty
+    - Vout:
+      specifies the list of new UTXOs;
+      **must** be nonempty
+
+## Transaction Validation
+
+For a transaction to be valid, it must
+ *  have valid type
+ *  have valid data corresponding to specified type
+ *  have at least one TXIN
+ *  have at least one UTXO
+ *  have valid transaction indices on all new UTXOs
+ *  have valid DataStore indices
+ *  have valid inputs (unique)
+ *  have matching input and output values
+ *  have valid fees
+ *  have valid TxHash
+
+Before a transaction is processed,
+all of its TxIn values must have valid signatures.
+
+## Merkle Trees
+
+We now discuss
+[Merkle trees](https://en.wikipedia.org/wiki/Merkle_tree)
+before we talk about transaction hashing.
+Our discussion is based on the
+[OpenZeppelin Merkle tree library](https://github.com/OpenZeppelin/merkle-tree).
+
+We can compute the Merkle Tree root hash using
+the following functions:
+
+```
+def LeafHash(data):
+    return Hash(Hash(data))
+
+def HairPair(a, b):
```

Comment:
def hair_pair(a, b):
    if int(a, 16) < int(b, 16):    # assuming a and b are hexadecimal strings
        return hashlib.sha256((a + b).encode()).hexdigest()
    else:
        return hashlib.sha256((b + a).encode()).hexdigest()